### PR TITLE
feat: add PartOptional

### PR DESCRIPTION
**Pull request by ShanaMaid**
## Related issues:
- Resolved #XXX

## Checklist

* [ ] I have read [CONTRIBUTING.md](../CONTRIBUTING.md)
* [ ] I have linked all related issues above
* [ ] I have rebased my branch

For bugfixes:
* [ ] I have added at least one unit test to confirm the bug have been fixed
* [ ] I have checked and updated TOC and API Docs when necessary

For new features:
* [x] I have added entry in TOC and API Docs
* [x] I have added a short example in API Docs to demonstrate new usage
* [x] I have added type unit tests with `dts-jest`

 I add `PartOptional `

image

### CODE REVIEW

**Comment by piotrwitek** [comment]
That's nice, but why not just use `Optional<T, K = any>` so it would work like that:
```ts
type Person = Optional<Example>; // all props are optional

type Person = Optional<Example, 'age' | 'height'>; // only selected keys are optional
```

**Comment by ShanaMaid** [comment]
@piotrwitek oh, you are right!I did it， please review once ?

File: src/mapped-types.ts [hunk 1]
```diff
@@ -554,3 +554,23 @@ type IfEquals<X, Y, A = X, B = never> = (<T>() => T extends X
...
+ *      height: number;
+ *    };
+ *
+ *    // Expect: { name?: string; age?: string; height?: number; }
+ *    type Props = Optional<Props>;
+ *
+ *    // Expect: { name: string; age?: string; height?: number; }
+ *    type Props = Optional<Props, 'age' | 'height'>;
+ */
+export type Optional<T extends {}, K = keyof any> = K extends (keyof T)
```
Comment (piotrwitek) [inline src/mapped-types.ts]: Please change `{}` to `object` as this is a convention used in this project

Comment (ShanaMaid) [reply src/mapped-types.ts]: done

**Comment by ShanaMaid** [comment]
@piotrwitek maybe all is ready?

**Comment by piotrwitek** [comment]
@ShanaMaid awesome job, thanks!
Before I can merge I need to investigate because resulting test type is weird, maybe it can be improved
```ts
(Pick<Props, "name" | "visible"> & { age?: number | undefined; }) | (Pick<Props, "name" | "age"> & { visible?: boolean | undefined; })
```

I'll take a look in 1 hour

**Comment by piotrwitek** [comment]
@ShanaMaid I have simplified the Optional type, now it has much clearer resulting type:
```ts
export type Optional<T extends object, K extends keyof T = keyof T> = 
Omit<T,K> & Partial<Pick<T, K>>;
```

I'll be releasing in v3.7

**Comment by ShanaMaid** [comment]
ok, that's cool!

**Comment by felixfbecker** [comment]
Can we get the inverse to, i.e. `Required<T, K>`?

**Comment by piotrwitek** [comment]
@felixfbecker definitely yes, as this type turned out to be quite useful.
Could you please create a new issue for it so myself or anyone else can pick it up?

**Pull request merged**